src/rax.c

```diff
@@ -499,6 +499,92 @@ static inline size_t raxLowWalk(rax *rax, unsigned char *s, size_t len, raxNode
     return i;
 }
 
+int raxIteratorPushChildOffset(raxIterator *it, uint8_t child_offset) {
```

Comment:
Why `child_offset` use `uint8_t`, `node->size` can be larger than 256.

Reply:
When node->size is used in a non-compressed node it cannot exceed 256.

Reply:
Oh - and if a node is compressed it can only have 1 child, hence an offset of 0.

Reply:
I think you misunderstand the way children's characters are stored, they are repeatable, not unique, so 256 is not enough.
.e.g an uncompress node with 1000 children

| iskey | isnull | iscompr | size  |  node1 ~ node1000 | data|
|  ----  | ----  | ----  |----  |----  |----  |
|    1bit  | 1bit  |   1bit     |  1000 | 1000 * sizeof(raxNode*) | sizeof(void*) |

Reply:
Actually, my reading of the code differs. Can you suggest an experiment which would allow us to use raxShow to illustrate your understanding?

Reply:
Here's gist of my version of raxShow ([raxShowHexKey](https://gist.github.com/michaelplaing/28934165071cb6cd04d99ec24eb4e858)) which is better for examining binary content and useful for examining experiments.

Reply:
Here's an experiment...

Let's start with a new tree:
```
rax* tree = raxNew();
```
and say we have only 3 possible characters in our domain: ``a, b, c`` and then do a few inserts:
```
raxInsert(tree, (uint8_t*)"c", 1, NULL, NULL);
raxInsert(tree, (uint8_t*)"a", 1, NULL, NULL);
raxInsert(tree, (uint8_t*)"b", 1, NULL, NULL);
```
now show the result:
```
raxShow(tree);
[abc]
 `-(a) []=0x0
 `-(b) []=0x0
 `-(c) []=0x0
```
The child keys are in sorted order regardless of insertion order.

Let's try to insert a second ``b`` to create a duplicate child key and show the result:
```
raxInsert(tree, (uint8_t*)"b", 1, NULL, NULL);
raxShow(tree);
[abc]
 `-(a) []=0x0
 `-(b) []=0x0
 `-(c) []=0x0
```
The tree did not change. The new ``b`` just replaced the old.

But let's try again – this time insert the key ``ba`` and show the result:
```
raxInsert(tree, (uint8_t*)"ba", 2, NULL, NULL);
raxShow(tree);
[abc]
 `-(a) []=0x0
 `-(b) [a]=0x0 -> []=0x0
 `-(c) []=0x0
```
The keys ``b`` and ``ba`` share the same unique character value in the root child key vector without duplication.

Let's expand the analysis in a new tree to the full range of 256 values for an unsigned character. We'll use a `uint16_t` and store it in big endian, as is done in Redis.

Insert the 1st 256 values. The order of insertion does not matter as the resulting child vector is always sorted in ascending order. Then use ``raxShowHexKey`` since ``raxShow`` is not very good with binary data.
```
uint16_t u16;
uint8_t u8v[2];
for (int i = 0; i < 256; i++) {
    u16 = i;
    u8v[0] = u16 >> 8; u8v[1] = u16 & 0xff; // big endian conversion
    raxInsert(tree, u8v, 2, NULL, NULL);
}
raxShowHexKey(tree);
[0x00]->[0x000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f  303132333435363738393a3b3c3d3e3f404142434445464748494a4b4c4d4e4f505152535455565758595a5b5c5d5e5f 606162636465666768696a6b6c6d6e6f707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f 909192939495969798999a9b9c9d9e9fa0a1a2a3a4a5a6a7a8a9aaabacadaeafb0b1b2b3b4b5b6b7b8b9babbbcbdbebf c0c1c2c3c4c5c6c7c8c9cacbcccdcecfd0d1d2d3d4d5d6d7d8d9dadbdcdddedfe0e1e2e3e4e5e6e7e8e9eaebecedeeef
f0f1f2f3f4f5f6f7f8f9fafbfcfdfeff]
       `—(.){[]}
       ...more unprintable chars
       `—(!){[]}
       ...more printable chars
       `—(~){[]}
       ...more unprintable chars
       `—(.){[]}
```
Trying to insert another integer in the range 0 to 255 will fail to make any difference in the tree, similar to above.

We can insert the integer 256, altering the tree:

```
u16 = 256;
u8v[0] = u16 >> 8; u8v[1] = u16 & 0xff; // big endian conversion
raxInsert(tree, u8v, 2, NULL, NULL);
raxShowHexKey(tree);
[0x0001]
 `—(.)[0x000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f
       303132333435363738393a3b3c3d3e3f404142434445464748494a4b4c4d4e4f505152535455565758595a5b5c5d5e5f
       606162636465666768696a6b6c6d6e6f707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f
       909192939495969798999a9b9c9d9e9fa0a1a2a3a4a5a6a7a8a9aaabacadaeafb0b1b2b3b4b5b6b7b8b9babbbcbdbebf
       c0c1c2c3c4c5c6c7c8c9cacbcccdcecfd0d1d2d3d4d5d6d7d8d9dadbdcdddedfe0e1e2e3e4e5e6e7e8e9eaebecedeeef
       f0f1f2f3f4f5f6f7f8f9fafbfcfdfeff]
       `—(.){[]}
       ...more unprintable chars
       `—(!){[]}
       ...more printable chars
       `—(~){[]}
       ...more unprintable chars
       `—(.){[]}
 `—(.)[0x00]->{[]}
```
Again there is no duplication of child keys, all of which are unique.

There is in fact no Rax operation that will create a duplicate Rax key or a duplicate child key.

Hence the maximum number of child keys at each level is 256, since there are 256 unique states for an 8-bit character and rax is strictly character based.

Reply:
Oh, you are right, I misunderstood the structure of compression and non-compression.

Reply:
It has taken me awhile to wrap my head around it!